DDInstall/DDInstall.psm1

```diff
@@ -0,0 +1,192 @@
+
+    .PARAMETER DDAgentUsername
+    Override the default ddagentuser username used during Agent installation (v6.11.0+).
+
+    .PARAMETER DDAgentPassword
+    Override the cryptographically secure password generated for the ddagentuser user during Agent installation (v6.11.0+). Must be provided for installs on domain servers.
+
+    .PARAMETER ApplicationDataDirectory
+    Override the directory to use for the configuration file directory tree (v6.11.0+). May only be provided on initial install; not valid for upgrades. Default: C:\ProgramData\Datadog
+
+    .PARAMETER ProjectLocation
+    Override the directory to use for the binary file directory tree (v6.11.0+). May only be provided on initial install; not valid for upgrades. Default: %ProgramFiles%\Datadog\Datadog Agent
+
+    .PARAMETER AgentInstallURL
+    Override the URL which the Agent installer is downloaded from.
+
+    .PARAMETER InstallLogPath
+    Override the Agent installation log location.
+
+    .INPUTS
+    None. You cannot pipe objects to Install-DDAgent.
+
+    .OUTPUTS
+    None.
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>"
+
+    .EXAMPLE
+    C:\PS> Install-DDAgent -ApiKey "<YOUR_DATADOG_API_KEY>" -WithAPMTracers "DotNet"
+
+    .LINK
+    Learn more about the Datadog Windows Agent User:
+    https://docs.datadoghq.com/agent/guide/windows-agent-ddagent-user/
+    #>
+}
+
+###############################
+# Unexported helper functions #
+###############################
+
+function formatAgentInstallerParameters($params)
+{
+    $formattedparams = "APIKEY=$($params.ApiKey)"
+
+    if ($params.ContainsKey('Site'))                        { $formattedparams += " SITE=$($params.Site)"}
+    if ($params.ContainsKey('Tags'))                        { $formattedparams += " TAGS=``$($params.Tags)``" }
+    if ($params.ContainsKey('Hostname'))                    { $formattedparams += " HOSTNAME=$($params.Hostname)" }
+    if ($params.ContainsKey('DDAgentUsername'))             { $formattedparams += " DDAGENTUSER_NAME=$($params.DDAgentUsername)" }
+    if ($params.ContainsKey('DDAgentPassword'))             { $formattedparams += " DDAGENTUSER_PASSWORD=$($params.DDAgentPassword)" }
+    if ($params.ContainsKey('ApplicationDataDirectory'))    { $formattedparams += " APPLICATIONDATADIRECTORY=$($params.ApplicationDataDirectory)" }
+    if ($params.ContainsKey('ProjectLocation'))             { $formattedparams += " PROJECTLOCATION=$($params.ProjectLocation)" }
+    if ($params.ContainsKey('InstallLogPath'))              { $formattedparams += " /log $($params.InstallLogPath)" }
+
+    return $formattedparams
+}
+
+function downloadAsset($url, $outFile)
+{
+    $ProgressPreference = 'SilentlyContinue'
+
+    Invoke-WebRequest -Uri $url -OutFile $outFile
+    # Suppress progress bar to prevent performance issues (affects older windows versions, see https://github.com/PowerShell/PowerShell/issues/13414)
```

Comment:
should this comment be up a few lines? Also I guess this means there wasn't a way to detect if we were running interactively or not :/

Reply:
ah yep, copy-paste error 😬

I found a couple recommendations of ways to check if the session was interactive, but I wasn't sure how reliable they were & figured it might just be safer to always disable it. But, if you think that showing a download progress bar for interactive sessions is worth doing, I can definitely add it & just spend some time testing it to make sure it's reliable!

Reply:
>  if you think that showing a download progress bar for interactive sessions is worth doing
I feel pretty strong that it's worth doing - when we were debugging this with a customer, it took a _long_ time for everything to complete and we were left wondering if the script hung or not.

Do you know what the performance impact is?